### ISSUE TITLE
Reset optimizer variables and learning rate

### ISSUE THREAD

**Issue by guillaumekln** [OpenNMT/OpenNMT-tf#229]
For a retraining, we sometimes want to restart with a "fresh" optimizer configuration. We should add a `SessionRunHook` that can:

* Override the global step to a specific value (e.g. 0 to reset the learning rate decay)
* Re-initialize the optimizer variables

Enabling these options should be done with command line arguments and not via the configuration file.

**Comment by atebbifakhr** [comment]
Hi,
I'm interested to contribute on this feature.
The problem is by reseting the global step the checkpoints will be overwritten, do you have any suggestion?
cheers

**Comment by guillaumekln** [comment]
Hi,

Thank you for your interest. You are right about the problem, I thought about this issue only after opening this thread.

One solution would be to not override the global step but introduce an offset. I don't like this one as it would make the code more complex, require a new variable in the model, and make it generally confusing for the user.

Another solution could be to allow starting the training from an existing directory, say `A`, that is is different than the configured `model_dir`, say `B`. Indeed, it makes sense (and is even better) for a retraining to not use the existing directory. This would typically require to:

* copy the [last checkpoint](https://www.tensorflow.org/api_docs/python/tf/train/latest_checkpoint) from `A` to `B`
* [generate a new checkpoint state](https://www.tensorflow.org/api_docs/python/tf/train/generate_checkpoint_state_proto) in `B`
* start the training (by default, the last checkpoint from `B` is loaded)
* [remove](https://www.tensorflow.org/api_docs/python/tf/train/remove_checkpoint) the copied checkpoint from `B` after the weights are loaded

(I think all this can be implemented in a single hook.)

What do you think? Let me know if you also want to work on this, it's independent from this issue.

**Comment by atebbifakhr** [comment]
Hi,

Yes, it makes sense. I was also thinking to change the directory.
I'll work on it asap, since I need this functionality.

Cheers

**Comment by atebbifakhr** [comment]
Hi again,

I finally ended up to the idea that is starting a new training from scratch for fine-tuning, and passing a checkpoint from previous step of training to load model weights from it.
By doing this, there is no need for copying and removing checkpoints.

I add 'LoadCheckpointHook' in order to load model weights from the checkpoint, which is passed in command line arguments.
